### PR TITLE
kernel: Adapt to low version Android init process

### PR DESCRIPTION
**Pull request by longhuan1999**
1. Adapt to low version Android init process
2. Add stop hook output
3. Fix output with missing line breaks

### CODE REVIEW

File: kernel/ksud.c [hunk 1]
```diff
@@ -103,6 +103,44 @@ static const char __user *get_user_arg_ptr(struct user_arg_ptr argv, int nr)
...
+        
+        // Replace equal sign with string terminator
+        *env_value = '\0';
+        env_value++;
+        
+        // pr_info("env %s=%s\n", env_name, env_value);
+        // Check if the environment variable names are matching
+        if (strcmp(env_name, name) == 0) {
+            // If matched, returns a Boolean value based on the environment variable's value
+            if (strcmp(env_value, "1") == 0 || strcmp(env_value, "true") == 0)
```
Comment (tiann) [inline kernel/ksud.c:132]:
```suggestion
return strcmp(env_value, "1") == 0 || strcmp(env_value, "true") == 0;
```

File: kernel/ksud.c [hunk 2]
```diff
@@ -163,24 +205,37 @@ int ksu_handle_execveat_ksud(int *fd, struct filename **filename_ptr,
 	}
 
 	if (unlikely(!memcmp(filename->name, system_bin_init,
-		    sizeof(system_bin_init) - 1))) {
```
Comment (tiann) [inline kernel/ksud.c:166]: Don't bother the logic of Android 10+. We don't want to introduce new bugs for just adapting Android 10-

Comment (tiann) [reply kernel/ksud.c:166]: You should make it a single condition block, don't modify code of Android 10+. If Android 6-7 needs to modify it, please drop.

**Review by tiann — CHANGES_REQUESTED** [review]

**Review by tiann — CHANGES_REQUESTED** [review]

File: kernel/ksud.c [hunk 3]
```diff
@@ -103,6 +103,41 @@ static const char __user *get_user_arg_ptr(struct user_arg_ptr argv, int nr)
 	return native;
 }
 
+static const bool get_env_bool(struct user_arg_ptr envp, const char *name) {
```
[the remark this answers is not in the document]
Comment (longhuan1999) [reply kernel/ksud.c:106]: I think so

[the remark this answers is not in the document]
Comment (tiann) [reply kernel/ksud.c:106]: So can we just use `get_user_arg_ptr`, i think it is more stable than yours.

[the remark this answers is not in the document]
Comment (longhuan1999) [reply kernel/ksud.c:106]: Actually, I don't know how to do it. I have not studied C++ and kernel development

[the remark this answers is not in the document]
Comment (tiann) [reply kernel/ksud.c:106]: 1. First, you can use `count(*envp, MAX_ARG_STRINGS)` to get the env count.
2. Second, you can use `get_user_arg_ptr(envp, i)` to get every env variable.
3. And then you can check every env variable to see if it hit your condition.

[the remark this answers is not in the document]
Comment (longhuan1999) [reply kernel/ksud.c]: ok

File: kernel/ksud.c [hunk 4]
```diff
@@ -183,10 +187,63 @@ int ksu_handle_execveat_ksud(int *fd, struct filename **filename_ptr,
...
+			}
+		} else if (argc == 1 && !init_second_stage_executed) {
+			/* This applies to versions between Android 8 ~ 9  */
+			int envc = count(*envp, MAX_ARG_STRINGS);
+			if (envc > 0) {
+				int n;
+				for (n = 1; n <= envc; n++) {
+					const char __user *p = get_user_arg_ptr(*envp, n);
+					if (p && !IS_ERR(p)) {
+						char env[256];
```
Comment (tiann) [inline kernel/ksud.c:219]: Why copy twice?

File: kernel/ksud.c [hunk 5]
```diff
@@ -183,10 +187,63 @@ int ksu_handle_execveat_ksud(int *fd, struct filename **filename_ptr,
...
+				pr_err("/init parse args err!\n");
+			}
+		} else if (argc == 1 && !init_second_stage_executed) {
+			/* This applies to versions between Android 8 ~ 9  */
+			int envc = count(*envp, MAX_ARG_STRINGS);
+			if (envc > 0) {
+				int n;
+				for (n = 1; n <= envc; n++) {
+					const char __user *p = get_user_arg_ptr(*envp, n);
+					if (p && !IS_ERR(p)) {
```
Comment (tiann) [inline kernel/ksud.c:218]:
```suggestion
					if (!p || IS_ERR(p)) {
					    continue;
					}
```

**Review by tiann — CHANGES_REQUESTED** [review]

File: kernel/ksud.c [hunk 6]
```diff
@@ -183,10 +187,65 @@ int ksu_handle_execveat_ksud(int *fd, struct filename **filename_ptr,
...
+			/* This applies to versions between Android 8 ~ 9  */
+			int envc = count(*envp, MAX_ARG_STRINGS);
+			if (envc > 0) {
+				int n;
+				for (n = 1; n <= envc; n++) {
+					const char __user *p = get_user_arg_ptr(*envp, n);
+					if (!p || IS_ERR(p)) {
+					    continue;
+					}
+					if (p && !IS_ERR(p)) {
```
Comment (tiann) [inline kernel/ksud.c:221]: we can delete this condition now.

**Review by tiann — CHANGES_REQUESTED** [review]

File: kernel/ksud.c (hunk 6 above)
Comment (longhuan1999) [reply kernel/ksud.c:221]: No need (p && !IS_ERR(p)) and (!p || IS_ERR(p))? 
What do they mean? 
Are they same with (ksu_strncpy_from_user_nofault(env, p, sizeof(env)) < 0)?

Comment (tiann) [reply kernel/ksud.c]: Because we have
```
if (!p || IS_ERR(p)) {
	continue;
}
```

Comment (longhuan1999) [reply kernel/ksud.c]: Is that all right now?

Comment (tiann) [reply kernel/ksud.c]: It looks good to me now. One final question: Does it work on your device? I have no device to test.

Comment (longhuan1999) [reply kernel/ksud.c]: I just tested it and it works properly on my device.

**Review by tiann — APPROVED** [review]

**Pull request merged**